### DISCUSSION TITLE
sendMessage must be string not list

### DISCUSSION THREAD

**Discussion by aggr0lite** [AsamK/signal-cli#1302] [Q&A]
Hi!

I am running signal-cli on system daemon and most of it works pretty well.
However, sometimes, it complains with the source parameter of the sendMessage method.

When i send it like this, it sometimes throws the error that it "must be string not list":
```python
signal.sendMessage("Service started",[],[ADMIN])
```

Then I changed the source parameter from list to string, which worked for some time but stopped working without any error again:
```python
signal.sendMessage("Service started",[],ADMIN)
```

I then changed it again from a string to a list which worked. 

Am I missing something? I don't really understand why it sometimes just stops working with the given error in the title and sometimes it works.

Edit:
Just got this situation again. Started a script on boot via systemd which uses the dbus. Service didn't start because it fails sending a message with this error:

`Traceback (most recent call last):
  File "/home/pi/signal-receive.py", line 235, in <module>
    signal.sendMessage("Service started",[],[ADMIN])
  File "/home/pi/.local/lib/python3.9/site-packages/pydbus/proxy_method.py", line 74, in __call__
    self._iface_name, self.__name__, GLib.Variant(self._sinargs, args), GLib.VariantType.new(self._soutargs),
  File "/usr/lib/python3/dist-packages/gi/overrides/GLib.py", line 189, in __new__
    v = creator._create(format_string, value)
  File "/usr/lib/python3/dist-packages/gi/overrides/GLib.py", line 150, in _create
    builder.add_value(self._create(dup, i))
  File "/usr/lib/python3/dist-packages/gi/overrides/GLib.py", line 118, in _create
    return self._LEAF_CONSTRUCTORS[format](value)
TypeError: Must be string, not list`

ADMIN = a phone number like '+491750000000'

The script which starts on boot and which fails with sendMessage looks like this:

```python
bus = SystemBus()
signal = bus.get('org.asamk.Signal') 
signalConfig = configparser.ConfigParser()
signalConfig.read(configPath)
ADMIN = signalConfig["admin"]["admin"]
signal.sendMessage("Service started",[],[ADMIN])
```
I run into this issue after doing a `sudo reboot`. Even manually executing the python script which should start on boot throws this error now. It worked well before I did the reboot. 

Any idea what's going on?

**Comment by greenpines65** [comment]
Hello,
Same thing had happened on my setup & **I believe this is a serious bug**.

I'm using precompiled signal-cli, Docker, Alpine Linux, and Pydbus or Dasbus to access the D-Bus interface.

The root cause of the problem: signal-cli publishes 2 D-Bus methods with the same name: `sendMessage(sasas)` and `sendMessage(sass)`:
```
[dbus-init-+222222222] TRACE org.freedesktop.dbus.MethodTuple - new MethodTuple(sendMessage, sass)
[dbus-init-+222222222] TRACE org.freedesktop.dbus.MethodTuple - new MethodTuple(sendMessage, sasas)
```

The Python library doesn't care which method to use - and creates the proxy objects for one or the other, randomly.

A careless user (me) calls it. 
```python
def onMessageReceived(self, timestamp, source, group_id, message, attachments):
    self.signal.sendMessage('ping', [], source)
```
If the proxy object is bound to the `sendMessage(sass)` variant, everything's OK. (And it had been OK for a while!)
Then the random choise reverses. The receiver doesn't see my messages. I look at the logs and freak out. **The signal-cli tried to send my message to invalid numbers!**
```
[DBusConnection [listener=false]] TRACE org.freedesktop.dbus.MethodTuple - new MethodTuple(sendMessage, sasas)
[RcvSvc: false DBus-MethodCall-Receiver-3] TRACE org.freedesktop.dbus.Marshalling - Deserializing from
    [ping, [], [+, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2]] to
    [class java.lang.String, java.util.List<java.lang.String>, java.util.List<java.lang.String>]
[RcvSvc: false DBus-MethodCall-Receiver-3] TRACE o.f.d.c.impl.DBusConnection - Invoking Method:
    public abstract long org.asamk.Signal.sendMessage(java.lang.String,java.util.List,java.util.List)
    throws org.asamk.Signal$Error$AttachmentInvalid,org.asamk.Signal$Error$Failure,org.asamk.Signal$Error$InvalidNumber,org.asamk.Signal$Error$UntrustedIdentity on org.asamk.signal.dbus.DbusSignalImpl@6b40c355
    with parameters [ping, [], [+, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2]]
...
[RcvSvc: false DBus-MethodCall-Receiver-3] WARN  LibSignal - [PhoneNumberFormatter]: Got local CC: ..
[RcvSvc: false DBus-MethodCall-Receiver-3] DEBUG LibSignal - [PhoneNumberFormatter]: NumberParseException: The string supplied did not seem to be a phone number.
[RcvSvc: false DBus-MethodCall-Receiver-3] WARN  LibSignal - [PhoneNumberFormatter]: Got local CC: ..
[RcvSvc: false DBus-MethodCall-Receiver-3] DEBUG LibSignal - [PhoneNumberFormatter]: NumberParseException: The string supplied did not seem to be a phone number.
[RcvSvc: false DBus-MethodCall-Receiver-3] DEBUG o.a.s.m.api.RecipientIdentifier - Normalized number 1 to +222222221.
[RcvSvc: false DBus-MethodCall-Receiver-3] WARN  LibSignal - [PhoneNumberFormatter]: Got local CC: ..
[RcvSvc: false DBus-MethodCall-Receiver-3] DEBUG LibSignal - [PhoneNumberFormatter]: NumberParseException: The string supplied did not seem to be a phone number.
[RcvSvc: false DBus-MethodCall-Receiver-3] DEBUG o.a.s.m.api.RecipientIdentifier - Normalized number 2 to +222222222.
[RcvSvc: false DBus-MethodCall-Receiver-3] WARN  LibSignal - [PhoneNumberFormatter]: Got local CC: ..
[RcvSvc: false DBus-MethodCall-Receiver-3] DEBUG LibSignal - [PhoneNumberFormatter]: NumberParseException: The string supplied did not seem to be a phone number.
[RcvSvc: false DBus-MethodCall-Receiver-3] DEBUG o.a.s.m.api.RecipientIdentifier - Normalized number 3 to +222222223.
...
[RcvSvc: false DBus-MethodCall-Receiver-4] TRACE o.a.s.manager.helper.RecipientHelper - Querying CDSI for 1 new numbers (0 previous), isPartialRefresh=true
[RcvSvc: false DBus-MethodCall-Receiver-4] DEBUG o.a.s.m.s.recipients.RecipientStore - Marking 1 numbers as unregistered
[RcvSvc: false DBus-MethodCall-Receiver-4] TRACE o.a.s.manager.helper.RecipientHelper - Querying CDSI for 1 new numbers (0 previous), isPartialRefresh=true
[RcvSvc: false DBus-MethodCall-Receiver-4] DEBUG o.a.s.m.s.recipients.RecipientStore - Marking 1 numbers as unregistered
...
[RcvSvc: false DBus-MethodCall-Receiver-4] TRACE o.a.s.manager.helper.RecipientHelper - Querying CDSI for 1 new numbers (0 previous), isPartialRefresh=true
java.net.ProtocolException: Expected HTTP 101 response but was '429 Too Many Requests'
...
[DBus Sender Thread: false, 1] TRACE o.f.d.s.m.OutputStreamMessageWriter - Message sent: MethodReturn(0,12)
    { Invalid=>null, Path=>null, Interface=>null, Member=>null, Error Name=>null, Reply Serial=>6,
        Destination=>:1.1, Sender=>null, Signature=>x, Unix FD=>null }
    { 1711111111111 }
```
(Fortunately for me, those numbers weren't registered in the Signal service. 😌)

------

As pointed out in LEW21/pydbus#19, referring https://dbus.freedesktop.org/doc/dbus-java/dbus-java/dbus-javase1.html: D-Bus does not support method overloading, only one method or signal should exist with each name.

I suspect that renaming all the duplicate methods in signal-cli daemon is the right way to resolve.
I don't expect a fix in pydbus & dasbus.

Meanwhile, my dirty hack looks like:
```python
def sendSignalMessage(self, message, attachments, dest):
    try:
        ts = self.signal.sendMessage(message, attachments, [dest]) # BUG in signal-cli: 2 D-Bus methods with the same name.
    except TypeError as e:
        ts = self.signal.sendMessage(message, attachments, dest)
```

Cheers,
--gp65

------

P.S. This is my setup:
```
$ uname -srv
Linux 5.15.146.1-microsoft-standard-WSL2 #1 SMP Thu Jan 11 04:09:03 UTC 2024

$ docker --version
Docker version 24.0.5, build 24.0.5-0ubuntu1~22.04.1

$ docker image list alpine
REPOSITORY   TAG       IMAGE ID       CREATED        SIZE
alpine       latest    05455a08881e   2 months ago

container # apk list --installed py3-dasbus signal-cli '*jdk*'
openjdk21-jre-headless-21.0.2_p13-r0 x86_64 {openjdk21} (GPL-2.0-with-classpath-exception) [installed]
py3-dasbus-1.7-r2 x86_64 {py3-dasbus} (LGPL-2.1-or-later) [installed]
signal-cli-0.13.0-r0 x86_64 {signal-cli} (GPL-3.0-or-later) [installed]
```

**Comment by greenpines65** [comment]
Other signal-cli issues that match the topic of calling overloaded D-Bus methods using pydbus:
#479 [DBus] send message to multiple number
#787 signal.sendMessage hangs
#807 [dbus] sendMessage recipient parameter
#903 isRegistered takes 0 positional argument
#980 [ERROR] sendMessageReaction

**Comment by Joost2020** [comment]
I have come across the same issue.
It seems to be related to the python version / install you are using.
I am on Ubuntu 24.04 and after a regular update (sudo apt upgrade) which also upgraded some python modules, this problem occurred.
On my test-server which did not get the upgrade yet I can freely use `<s>` and `<as>`, string and array of strings.
On my production server I can now only use `<s> `string. Using an array of strings gives a typeError.

Moreover I have to explicitly typeCast it as a dbus.string. It does not accept regular Python strings anymore, whereas with the older python this was not a problem.

Edit:
Sorry I did not read correctly. Previous was about SendMessage
I am talking about SendMessageReaction
But maybe to some extent this applies as well. So maybe I just leave this comment here. If someone thinks this should be deleted that's fine too.
